posts/c-abi-changes-for-wasm32-unknown-unknown.md

```diff
@@ -0,0 +1,263 @@
+where the WASI targets for Rust use a corrected ABI definition. Still, however,
+the non-standard definition remained in use for `wasm32-unknown-unknown`.
+
+The time has now come to correct this historical mistake and the Rust compiler
+will soon be using a correct ABI definition for the `wasm32-unknown-unknown`
+target. This means, however, that generated WebAssembly binaries will be
+different than before.
+
+## What is a WebAssembly C ABI?
+
+The definition of an ABI answers questions along the lines of:
+
+* What registers are arguments passed in?
+* What registers are results passed in?
+* How is a 128-bit integers passed as an argument?
+* How is a `union` passed as a return value?
+* When are parameters passed through memory instead of registers?
+* What is the size and alignment of a type in memory?
+
+For WebAssembly these answers are a little different than native platforms.
+WebAssembly does not have physical registers for example and functions must all
+be annotated with a type. What WebAssembly does have is types such as `i32`,
+`i64`, `f32`, and `f64`. This means that for WebAssembly an ABI needs to define
+how to represent values in these types.
+
+This is where the [tool-conventions] document comes in. That document provides a
+definition for how to represent primitives in C in the WebAssembly format, and
+additionally how function signatures in C are mapped to function signatures in
+WebAssembly. For example a Rust `u32` is represented by a WebAssembly `i32` and
+is passed directly as a parameter as a function argument. If the Rust structure
+`#[repr(C)] struct Pair(f32, f64)` is returned from a function then a return
+pointer is used which must have alignment 8 and size of 16 bytes.
+
+In essence the WebAssembly C ABI is acting as a bridge between C's type system
+and the WebAssembly type system. This includes details such as in-memory layouts
+and translations of a C function signature to a WebAssembly function signature.
+
+## How is `wasm32-unknown-unknown` non-standard?
+
+Despite the ABI definition today being non-standard many aspects of it are
+still the same as what [tool-conventions] specifies. For example size/alignment
+of types is the same as it is in C. The main difference is how function
+signatures are calculated. An example (where you can follow along on [godbolt])
+is:
+
+```rust
+#[repr(C)]
+pub struct Pair {
+    x: u32,
+    y: u32,
+}
+
+#[unsafe(no_mangle)]
+pub extern "C" fn pair_add(pair: Pair) -> u32 {
+    pair.x + pair.y
+}
+```
+
+This will generate the following WebAssembly function:
+
+```wasm
+(func $pair_add (param i32 i32) (result i32)
+  local.get 1
+  local.get 0
+  i32.add
+)
+```
+
+Notably you can see here that the struct `Pair` was "splatted" into its two
+components so the actual `$pair_add` function takes two arguments, the `x` and
+`y` fields. The [tool-conventions], however [specifically says] that "other
```

Comment:
`[specifically says]` doesn't link to anything.

Reply:
Oops thanks, I was originally going to link something here but realized that I couldn't get a more specific link so I inlined the words instead and forgot to remove the brackets!